File: solutions/python/exercise2/password_validator.py
from typing import Optional

MIN_LENGTH = 8

class PasswordValidator:
    def validate(self, password: Optional[str]) -> bool:
        # Exercise 2, Test: test_null_or_empty_returns_false
        # Reject None, empty, or whitespace-only input up front.
        if password is None or password.strip() == "":
            return False

        # Exercise 2, Test: test_less_than_eight_characters_returns_false
        # Enforce a minimum length.
        if len(password) < MIN_LENGTH:
            return False

        has_upper = any(c.isupper() for c in password)
        has_lower = any(c.islower() for c in password)
        has_digit = any(c.isdigit() for c in password)
        has_special = any(not c.isalnum() and not c.isspace() for c in password)

        # Exercise 2, Test: test_missing_uppercase_returns_false
        # Require an uppercase letter, a lowercase letter, and a special character
        # as part of the same "complex password" rule.
        if not (has_upper and has_lower and has_special):
            return False

        # Exercise 2, Test: test_missing_digit_returns_false
        # Require at least one digit.
        if not has_digit:
            return False

        # Exercise 2, Test: test_contains_sequential_digits_returns_false
        # Reject any run of three ascending sequential digits (e.g. "123").
        if _has_sequential_digits(password):
            return False

        # Exercise 2, Test: test_valid_complex_password_returns_true
        # Anything that clears all the checks above is a valid password.
        return True


def _has_sequential_digits(password: str) -> bool:
    for i in range(len(password) - 2):
        window = password[i:i + 3]
        if window.isdigit():
            a, b, c = (int(ch) for ch in window)
            if b == a + 1 and c == b + 1:
                return True
    return False

File: solutions/csharp/Exercise2/PasswordValidator.cs
namespace Workshop.Exercise2;

public interface IPasswordValidator
{
    bool Validate(string? password);
}

public class PasswordValidator : IPasswordValidator
{
    private const int MinLength = 8;

    public bool Validate(string? password)
    {
        // Exercise 2, Test: Validate_NullOrEmpty_ReturnsFalse
        // Reject null, empty, or whitespace-only input up front.
        if (string.IsNullOrWhiteSpace(password))
        {
            return false;
        }

        // Exercise 2, Test: Validate_LessThanEightCharacters_ReturnsFalse
        // Enforce a minimum length.
        if (password.Length < MinLength)
        {
            return false;
        }

        bool hasUpper = false, hasLower = false, hasDigit = false, hasSpecial = false;
        foreach (char c in password)
        {
            if (char.IsUpper(c)) hasUpper = true;
            else if (char.IsLower(c)) hasLower = true;
            else if (char.IsDigit(c)) hasDigit = true;
            else if (!char.IsWhiteSpace(c)) hasSpecial = true;
        }

        // Exercise 2, Test: Validate_MissingUppercase_ReturnsFalse
        // Require an uppercase letter, a lowercase letter, and a special character
        // as part of the same "complex password" rule.
        if (!hasUpper || !hasLower || !hasSpecial)
        {
            return false;
        }

        // Exercise 2, Test: Validate_MissingDigit_ReturnsFalse
        // Require at least one digit.
        if (!hasDigit)
        {
            return false;
        }

        // Exercise 2, Test: Validate_ContainsSequentialDigits_ReturnsFalse
        // Reject any run of three ascending sequential digits (e.g. "123").
        if (HasSequentialDigits(password))
        {
            return false;
        }

        // Exercise 2, Test: Validate_ValidComplexPassword_ReturnsTrue
        // Anything that clears all the checks above is a valid password.
        return true;
    }

    private static bool HasSequentialDigits(string password)
    {
        for (int i = 0; i < password.Length - 2; i++)
        {
            if (char.IsDigit(password[i]) && char.IsDigit(password[i + 1]) && char.IsDigit(password[i + 2]))
            {
                int a = password[i] - '0';
                int b = password[i + 1] - '0';
                int c = password[i + 2] - '0';
                if (b == a + 1 && c == b + 1)
                {
                    return true;
                }
            }
        }
        return false;
    }
}
